File: apps/backend/src/server/hooks.go
package server

import (
	"encoding/json"
	"log/slog"
	"net/http"

	"github.com/lucas/oraculo/apps/backend/src/db"
	"github.com/lucas/oraculo/apps/backend/src/ws"
)

// HookHandler handles inbound webhook calls from Claude Code hooks.
type HookHandler struct {
	agents   *db.AgentStore
	toolEvts *db.ToolEventStore
	sessEvts *db.SessionEventStore
	hub      *ws.Hub
	logger   *slog.Logger
}

func (h *HookHandler) broadcast(event string, payload any) {
	msg, err := json.Marshal(map[string]any{"event": event, "data": payload})
	if err != nil {
		return
	}
	h.hub.Broadcast(msg)
	h.logger.Info("hook.broadcast", "event", event)
}

// handleAgentStart records a new agent start event.
// POST /hooks/agent-start
// Body: {"session_id":"...","agent_name":"...","agent_type":"..."}
func (h *HookHandler) handleAgentStart(w http.ResponseWriter, r *http.Request) {
	var body struct {
		SessionID string `json:"session_id"`
		AgentName string `json:"agent_name"`
		AgentType string `json:"agent_type"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeAPIError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	h.logger.Info("hook.agent_started", "agent", body.AgentName, "type", body.AgentType)
	agent, err := h.agents.Start(body.SessionID, body.AgentName, body.AgentType)
	if err != nil {
		h.logger.Warn("hook.agent_started.error", "err", err)
		writeAPIError(w, http.StatusInternalServerError, err.Error())
		return
	}
	h.broadcast("agent_started", agent)
	writeJSON(w, agent)
}

// handleAgentStop records an agent stop event.
// POST /hooks/agent-stop
// Body: {"agent_id":1,"status":"completed"}
func (h *HookHandler) handleAgentStop(w http.ResponseWriter, r *http.Request) {
	var body struct {
		AgentID int    `json:"agent_id"`
		Status  string `json:"status"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeAPIError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	h.logger.Info("hook.agent_stopped", "agent_id", body.AgentID, "status", body.Status)
	agent, err := h.agents.Stop(body.AgentID, body.Status)
	if err != nil {
		h.logger.Warn("hook.agent_stopped.error", "err", err)
		writeAPIError(w, http.StatusInternalServerError, err.Error())
		return
	}
	h.broadcast("agent_stopped", agent)
	writeJSON(w, agent)
}

// handleToolUsed records a tool usage event.
// POST /hooks/tool-used
// Body: {"session_id":"...","tool_name":"...","file_path":"..."}
func (h *HookHandler) handleToolUsed(w http.ResponseWriter, r *http.Request) {
	var body struct {
		SessionID string `json:"session_id"`
		ToolName  string `json:"tool_name"`
		FilePath  string `json:"file_path"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeAPIError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	h.logger.Info("hook.tool_used", "tool", body.ToolName, "file", body.FilePath)
	if err := h.toolEvts.Record(body.SessionID, body.ToolName, body.FilePath); err != nil {
		h.logger.Warn("hook.tool_used.error", "err", err)
		writeAPIError(w, http.StatusInternalServerError, err.Error())
		return
	}
	h.broadcast("tool_used", body)
	writeJSON(w, map[string]string{"status": "ok"})
}

// handleTaskCompleted broadcasts a task completion event.
// POST /hooks/task-completed
// Body: {"session_id":"...","task_name":"...","status":"..."}
func (h *HookHandler) handleTaskCompleted(w http.ResponseWriter, r *http.Request) {
	var body struct {
		SessionID string `json:"session_id"`
		TaskName  string `json:"task_name"`
		Status    string `json:"status"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeAPIError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	h.logger.Info("hook.task_completed", "task", body.TaskName, "status", body.Status)

	payload, _ := json.Marshal(map[string]string{
		"task_name": body.TaskName,
		"status":    body.Status,
	})
	if err := h.sessEvts.Record(body.SessionID, "task_completed", string(payload)); err != nil {
		h.logger.Warn("hook.task_completed.record_error", "err", err)
	}

	h.broadcast("task_completed", body)
	writeJSON(w, map[string]string{"status": "ok"})
}

// handleStop broadcasts a server stop event.
// POST /hooks/stop
func (h *HookHandler) handleStop(w http.ResponseWriter, r *http.Request) {
	var body struct {
		SessionID            string `json:"session_id"`
		LastAssistantMessage string `json:"last_assistant_message"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeAPIError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	h.logger.Info("hook.stop", "session_id", body.SessionID)

	payload, _ := json.Marshal(map[string]string{
		"last_assistant_message": body.LastAssistantMessage,
	})
	if err := h.sessEvts.Record(body.SessionID, "stop", string(payload)); err != nil {
		h.logger.Warn("hook.stop.record_error", "err", err)
	}

	h.broadcast("stop", body)
	writeJSON(w, map[string]string{"status": "ok"})
}

// handleTeammateIdle broadcasts a teammate idle event.
// POST /hooks/teammate-idle
func (h *HookHandler) handleTeammateIdle(w http.ResponseWriter, r *http.Request) {
	var body struct {
		SessionID    string `json:"session_id"`
		TeammateName string `json:"teammate_name"`
		TeamName     string `json:"team_name"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeAPIError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	h.logger.Info("hook.teammate_idle", "teammate", body.TeammateName)

	payload, _ := json.Marshal(map[string]string{
		"teammate_name": body.TeammateName,
		"team_name":     body.TeamName,
	})
	if err := h.sessEvts.Record(body.SessionID, "teammate_idle", string(payload)); err != nil {
		h.logger.Warn("hook.teammate_idle.record_error", "err", err)
	}

	h.broadcast("teammate_idle", body)
	writeJSON(w, map[string]string{"status": "ok"})
}

// handleSessionStart broadcasts a session start event.
// POST /hooks/session-start
// Body: {"session_id":"..."}
func (h *HookHandler) handleSessionStart(w http.ResponseWriter, r *http.Request) {
	var body struct {
		SessionID string `json:"session_id"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeAPIError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	h.logger.Info("hook.session_started", "session_id", body.SessionID)
	h.broadcast("session_started", body)
	writeJSON(w, map[string]string{"status": "ok"})
}

// handleSessionEnd broadcasts a session end event.
// POST /hooks/session-end
// Body: {"session_id":"..."}
func (h *HookHandler) handleSessionEnd(w http.ResponseWriter, r *http.Request) {
	var body struct {
		SessionID string `json:"session_id"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeAPIError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	h.logger.Info("hook.session_ended", "session_id", body.SessionID)
	h.broadcast("session_ended", body)
	writeJSON(w, map[string]string{"status": "ok"})
}

File: apps/orchestrator/src/server/hooks.ts
import { Hono } from "hono";
import type { EpicStore } from "../db/stores/epic-store.ts";
import type { StoryStore } from "../db/stores/story-store.ts";
import type { TaskStore } from "../db/stores/task-store.ts";
import type { AgentStore } from "../db/stores/agent-store.ts";
import type { ToolEventStore } from "../db/stores/tool-event-store.ts";
import type { SessionEventStore } from "../db/stores/session-event-store.ts";
import type { WebSocketHub } from "./ws.ts";
import type { Broadcaster } from "./broadcaster.ts";

export interface HookDeps {
  epics: EpicStore;
  stories: StoryStore;
  tasks: TaskStore;
  agents: AgentStore;
  toolEvents: ToolEventStore;
  sessionEvents: SessionEventStore;
  hub: WebSocketHub;
  broadcaster: Broadcaster;
}

function broadcast(deps: HookDeps, event: string, payload: unknown) {
  deps.hub.broadcast({ event, data: payload });
  deps.broadcaster.push({ event, data: payload });
}

export function hookRoutes(deps: HookDeps): Hono {
  const app = new Hono();

  // POST /hooks/session-start
  app.post("/session-start", async (c) => {
    const body = await c.req.json<{ session_id?: string }>().catch(() => null);
    if (!body) return c.json({ error: "invalid JSON body" }, 400);

    broadcast(deps, "session_started", body);
    return c.json({ status: "ok" });
  });

  // POST /hooks/session-end
  app.post("/session-end", async (c) => {
    const body = await c.req.json<{ session_id?: string }>().catch(() => null);
    if (!body) return c.json({ error: "invalid JSON body" }, 400);

    broadcast(deps, "session_ended", body);
    return c.json({ status: "ok" });
  });

  // POST /hooks/agent-start
  app.post("/agent-start", async (c) => {
    const body = await c.req
      .json<{
        session_id?: string;
        agent_name?: string;
        agent_type?: string;
        task_name?: string;
        story_name?: string;
        epic_name?: string;
      }>()
      .catch(() => null);
    if (!body) return c.json({ error: "invalid JSON body" }, 400);

    if (!body.task_name || !body.story_name || !body.epic_name) {
      return c.json(
        { error: "task_name, story_name, and epic_name are required" },
        400,
      );
    }

    let epic;
    try {
      epic = deps.epics.getByName(body.epic_name);
    } catch {
      return c.json({ error: `epic not found: ${body.epic_name}` }, 404);
    }

    let story;
    try {
      story = deps.stories.getByName(epic.id, body.story_name);
    } catch {
      return c.json({ error: `story not found: ${body.story_name}` }, 404);
    }

    let task;
    try {
      task = deps.tasks.getByName(story.id, body.task_name);
    } catch {
      return c.json({ error: `task not found: ${body.task_name}` }, 404);
    }

    try {
      const agent = deps.agents.Create(
        body.session_id ?? "",
        body.agent_name ?? "",
        body.agent_type ?? "",
        task.id,
      );
      broadcast(deps, "agent_started", agent);
      return c.json(agent);
    } catch (err) {
      return c.json(
        { error: err instanceof Error ? err.message : String(err) },
        500,
      );
    }
  });

  // POST /hooks/agent-stop
  app.post("/agent-stop", async (c) => {
    const body = await c.req
      .json<{ agent_id?: number; status?: string }>()
      .catch(() => null);
    if (!body) return c.json({ error: "invalid JSON body" }, 400);

    try {
      deps.agents.Stop(body.agent_id ?? 0);
      broadcast(deps, "agent_stopped", body);
      return c.json({ status: "ok" });
    } catch (err) {
      return c.json(
        { error: err instanceof Error ? err.message : String(err) },
        500,
      );
    }
  });

  // POST /hooks/tool-used
  app.post("/tool-used", async (c) => {
    const body = await c.req
      .json<{
        session_id?: string;
        tool_name?: string;
        file_path?: string;
      }>()
      .catch(() => null);
    if (!body) return c.json({ error: "invalid JSON body" }, 400);

    try {
      deps.toolEvents.Create(
        body.session_id ?? "",
        body.tool_name ?? "",
        body.file_path ?? "",
      );
      broadcast(deps, "tool_used", body);
      return c.json({ status: "ok" });
    } catch (err) {
      return c.json(
        { error: err instanceof Error ? err.message : String(err) },
        500,
      );
    }
  });

  // POST /hooks/task-completed
  app.post("/task-completed", async (c) => {
    const body = await c.req
      .json<{
        session_id?: string;
        task_name?: string;
        status?: string;
      }>()
      .catch(() => null);
    if (!body) return c.json({ error: "invalid JSON body" }, 400);

    const payload = JSON.stringify({
      task_name: body.task_name ?? "",
      status: body.status ?? "",
    });

    try {
      deps.sessionEvents.Create(
        body.session_id ?? "",
        "task_completed",
        payload,
      );
    } catch {
      // Non-fatal: log recording failure
    }

    broadcast(deps, "task_completed", body);
    return c.json({ status: "ok" });
  });

  // POST /hooks/task-started
  app.post("/task-started", async (c) => {
    const body = await c.req
      .json<{
        session_id?: string;
        task_name?: string;
        story_name?: string;
        epic_name?: string;
      }>()
      .catch(() => null);
    if (!body) return c.json({ error: "invalid JSON body" }, 400);

    if (!body.task_name || !body.story_name || !body.epic_name) {
      return c.json(
        { error: "task_name, story_name, and epic_name are required" },
        400,
      );
    }

    broadcast(deps, "task_started", body);
    return c.json({ status: "ok" });
  });

  // POST /hooks/stop
  app.post("/stop", async (c) => {
    const body = await c.req
      .json<{
        session_id?: string;
        last_assistant_message?: string;
      }>()
      .catch(() => null);
    if (!body) return c.json({ error: "invalid JSON body" }, 400);

    const payload = JSON.stringify({
      last_assistant_message: body.last_assistant_message ?? "",
    });

    try {
      deps.sessionEvents.Create(body.session_id ?? "", "stop", payload);
    } catch {
      // Non-fatal
    }

    broadcast(deps, "stop", body);
    return c.json({ status: "ok" });
  });

  // POST /hooks/teammate-idle
  app.post("/teammate-idle", async (c) => {
    const body = await c.req
      .json<{
        session_id?: string;
        teammate_name?: string;
        team_name?: string;
      }>()
      .catch(() => null);
    if (!body) return c.json({ error: "invalid JSON body" }, 400);

    const payload = JSON.stringify({
      teammate_name: body.teammate_name ?? "",
      team_name: body.team_name ?? "",
    });

    try {
      deps.sessionEvents.Create(
        body.session_id ?? "",
        "teammate_idle",
        payload,
      );
    } catch {
      // Non-fatal
    }

    broadcast(deps, "teammate_idle", body);
    return c.json({ status: "ok" });
  });

  return app;
}
